core/ecschema-metadata/src/Metadata/Schema.ts

```diff
@@ -443,29 +443,182 @@ export class Schema implements CustomAttributeContainerProps {
    * @param key the local (unqualified) name, lookup is case-insensitive
+   * @param itemConstructor The constructor of the item to return.
+   */
+  public getItemSync(name: string): SchemaItem | undefined
+  public getItemSync<T extends typeof SchemaItem>(name: string, itemConstructor: T): InstanceType<T> | undefined
+  public getItemSync<T extends typeof SchemaItem>(name: string, itemConstructor?: T): SchemaItem | InstanceType<T> | undefined {
+    const value = this._items.get(name.toUpperCase());
+    if (value === undefined || itemConstructor === undefined)
+      return value;
+
+    if (value.schemaItemType !== itemConstructor.schemaItemType) {
+      // There is one special case here: ECClass, where the item type can be any of the class types
+      if(itemConstructor.schemaItemType === AbstractSchemaItemType.Class && (
+        value.schemaItemType === SchemaItemType.EntityClass ||
+        value.schemaItemType === SchemaItemType.Mixin ||
+        value.schemaItemType === SchemaItemType.StructClass ||
+        value.schemaItemType === SchemaItemType.CustomAttributeClass ||
+        value.schemaItemType === SchemaItemType.RelationshipClass)) {
+          return value as InstanceType<T>;
+      }
+
+      if(itemConstructor.schemaItemType === AbstractSchemaItemType.SchemaItem) {
+        return value as InstanceType<T>;
+      }
+
+      return undefined;
+    }
+
+
+    return value as InstanceType<T>;
+  }
+
+  /**
+   * Attempts to find a schema item within this schema or a (directly) referenced schema
+   * @param key The full name or a SchemaItemKey identifying the desired item.
    */
-  public getItemSync<T extends SchemaItem>(name: string): T | undefined {
-    // Case-insensitive search
-    return this._items.get(name.toUpperCase()) as T;
+  public async lookupItem(key: Readonly<SchemaItemKey> | string): Promise<SchemaItem | undefined> {
+    let schemaName, itemName: string;
+    if (typeof (key) === "string") {
+      [schemaName, itemName] = SchemaItem.parseFullName(key);
+    } else {
+      itemName = key.name;
+      schemaName = key.schemaName;
+    }
+
+    if (!schemaName || schemaName.toUpperCase() === this.name.toUpperCase()) {
+      return this.getItem(itemName);
+    }
+
+    const refSchema = await this.getReference(schemaName);
+    if (!refSchema)
+      return undefined;
+
+    return refSchema.getItem(itemName);
   }
 
   /**
    * Attempts to find a schema item within this schema or a (directly) referenced schema
    * @param key The full name or a SchemaItemKey identifying the desired item.
+   * @param itemConstructor The constructor of the item to return.
    */
-  public async lookupItem<T extends SchemaItem>(key: Readonly<SchemaItemKey> | string): Promise<T | undefined> {
+  public async lookupTypedItem<T extends typeof SchemaItem>(key: Readonly<SchemaItemKey> | string, itemConstructor: T): Promise<InstanceType<T> | undefined> {
```

Comment:
After we unified _getItem_ in one method with multiple overloads, we should probably do the same with lookupType either.